src/pytest_recording/network.py

```diff
@@ -75,34 +94,46 @@ def uninstall_pycurl_wrapper():
     sys.modules["pycurl"] = pycurl
 
 
-def block_pycurl():
+def block_pycurl(allowed_hosts=None):
     global _disable_pycurl  # pylint: disable=global-statement
+    global _allowed_hosts  # pylint: disable=global-statement
     _disable_pycurl = True
+    _allowed_hosts = allowed_hosts
 
 
 def unblock_pycurl():
     global _disable_pycurl  # pylint: disable=global-statement
+    global _allowed_hosts  # pylint: disable=global-statement
     _disable_pycurl = False
+    _allowed_hosts = None
 
 
-def block_socket():
-    socket.socket.connect = network_guard
-    socket.socket.connect_ex = network_guard
+def block_socket(allowed_hosts=None):
+    socket.socket.connect = make_network_guard(_original_connect, allowed_hosts=allowed_hosts)
+    socket.socket.connect_ex = make_network_guard(_original_connect_ex, allowed_hosts=allowed_hosts)
 
 
 def unblock_socket():
     socket.socket.connect = _original_connect
     socket.socket.connect_ex = _original_connect_ex
 
 
-def network_guard(*args, **kwargs):
-    raise RuntimeError("Network is disabled")
+def make_network_guard(original_func, allowed_hosts=None):
+    def network_guard(*args, **kwargs):
+        if allowed_hosts:
+            # Make a regex that matches if any of our regexes match.
+            combined = "(" + ")|(".join(allowed_hosts) + ")"
+            if re.match(combined, args[1][0]):
```

Comment:
I propose to change the signature - put `self` and `address` to make it more clear - afaik it is the same across all Python versions. 

Also, `address` format depends on the address family, it is a 2-elements tuple for `AF_INET`, but for `AF_UNIX` it is a string, In this case, probably we should treat it as a value for comparison with allowed hosts